### PR TITLE
feat(artifacts): Add storage policy while adding files to artifacts

### PR DESCRIPTION
**Pull request by ibindlish**
Description
-----------

- Fixes [WB-17872](https://wandb.atlassian.net/browse/WB-17872)

What does the PR do? Include a concise description of the PR contents.
Add a policy="immutable" option to Artifact.add_file and Artifact.add_dir that prevents creating copies as staging files before uploading. 

Note:
1. I haven't made any changes to the integrations that use the add_file() and add_dir() methods.

Testing
-------
`pytest_tests/system_tests/test_artifacts/test_wandb_artifacts_full.py`
`pytest_tests/unit_tests/test_artifacts/test_storage.py`

Tested to make sure nexus still works the same.
```
import wandb

import os
import random
import tempfile

os.environ["WANDB_BASE_URL"] = "https://api.wandb.ai"
# os.environ["WANDB_BASE_URL"] = "http://localhost:8080"
# os.environ["WANDB_BASE_URL"] = "http://api.wandb.test"
# os.environ["WANDB_BASE_URL"] = "https://api.qa.wandb.ai"
# os.environ["WANDB_BASE_URL"] = "https://qa-aws.wandb.io"

ENTITY = "ibindlish"
PROJECT = "test-staging"
ARTIFACT = "test-art1"

tmpdir = ".."
with wandb.init(entity=ENTITY, project=PROJECT) as run:
    artifact = wandb.Artifact(ARTIFACT, "dataset")
    for i in range(5):
        with tempfile.NamedTemporaryFile(delete=False) as temp_file:
            data_to_write = "Hello, World!\nThis is a new temporary file."
            temp_file.write(data_to_write.encode("utf-8"))
        entry = artifact.add_file(temp_file.name, policy="immutable")
    artifact = run.log_artifact(artifact)
```

[WB-17872]: https://wandb.atlassian.net/browse/WB-17872?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tests/pytest_tests/system_tests/test_artifacts/test_wandb_artifacts_full.py [hunk 1]
```diff
@@ -236,6 +236,39 @@ def dir_size():
     assert dir_size() == 0
 
 
+def test_mutable_uploads(tmp_path, monkeypatch):
```
Comment (ibindlish) [inline tests/pytest_tests/system_tests/test_artifacts/test_wandb_artifacts_full.py:239]: These tests will be updated in the follow up PR that makes caching optional https://github.com/wandb/wandb/pull/7098 to include combinations of policy selections with cache selections.

**Comment by moredatarequired** [comment]
Do you mind changing the base to #7098 to highlight the actual content of this PR?

**Comment by ibindlish** [comment]
> Do you mind changing the base to #7098 to highlight the actual content of this PR?

Oh, i did it the other way around. So this PR (only staging changes) is based off `master`
And then the caching stuff (#7098) is based off this one...

**Review by moredatarequired — APPROVED** [review]
Thanks!

File: wandb/sdk/artifacts/artifact.py [hunk 2]
```diff
@@ -1122,13 +1123,14 @@ def new_file(
...
 
-        self.add_file(path, name=name)
+        self.add_file(path, name=name, policy="immutable")
 
     def add_file(
         self,
         local_path: str,
         name: Optional[str] = None,
         is_tmp: Optional[bool] = False,
+        policy: Optional[str] = "mutable",
```
Comment (kptkin) [inline wandb/sdk/artifacts/artifact.py:1133]: could you make this one a Literal, also it doesn't seem to be optional.

File: wandb/sdk/artifacts/artifact.py [hunk 3]
```diff
@@ -1159,16 +1164,24 @@ def add_file(
...
 
-        return self._add_local_file(name, local_path, digest=digest)
+        return self._add_local_file(name, local_path, digest=digest, policy=policy)
 
-    def add_dir(self, local_path: str, name: Optional[str] = None) -> None:
+    def add_dir(
+        self,
+        local_path: str,
+        name: Optional[str] = None,
+        policy: Optional[str] = "mutable",
```
Comment (kptkin) [inline wandb/sdk/artifacts/artifact.py:1173]: same as above

File: wandb/sdk/artifacts/artifact.py [hunk 4]
```diff
@@ -1386,18 +1399,30 @@ def do_write(f: IO) -> None:
         return entry
 
     def _add_local_file(
-        self, name: StrPath, path: StrPath, digest: Optional[B64MD5] = None
+        self,
+        name: StrPath,
+        path: StrPath,
+        digest: Optional[B64MD5] = None,
+        policy: Optional[str] = "mutable",
```
Comment (kptkin) [inline wandb/sdk/artifacts/artifact.py:1406]: same

File: wandb/sdk/artifacts/artifact.py (hunk 2 above)
Comment (ibindlish) [reply wandb/sdk/artifacts/artifact.py:1133]: Or did you mean that a Literal[] cannot be optional?

File: wandb/sdk/artifacts/artifact.py [hunk 5]
```diff
@@ -1386,18 +1400,30 @@ def do_write(f: IO) -> None:
         return entry
 
     def _add_local_file(
-        self, name: StrPath, path: StrPath, digest: Optional[B64MD5] = None
+        self,
+        name: StrPath,
+        path: StrPath,
+        digest: Optional[B64MD5] = None,
+        policy: Optional[Literal["mutable", "immutable"]] = "mutable",
```
Comment (kptkin) [inline wandb/sdk/artifacts/artifact.py:1407]: based on your check, doesn't look like policy is Optional

Comment (ibindlish) [reply wandb/sdk/artifacts/artifact.py:1407]: fixed it, thanks for catching this

**Pull request merged**